---
title: "Analysis of BLAST Results"
author: "Chelsea Noack"
date: "October 11, 2017"
output: github_document
---

# Introduction

Microbial communities within the enviornment span from sizes of large to small and are often predicted by "pairwise interactions" between the different species within the community. As bacterial species interact with one another, they in turn affect the health of their host. In that same vein, forensic studies can be helped by further understanding bacterial-host interactions. Skin-associated bacterial communities are highly diverse, yet personalized, which indicates a possibility for their use in forensic studies. 

In a study done by Fierer et al., skin-associated bacterial communities were gathered from single computer keys and computer mice and hypothesized to determine who owns the objects based on the genetic structures of the bacteria and the individual. By using a high-throughput pyrosequencing technique, Fierer et al. argue that one can confidently compare bacterial communities found on an individual's skin to the objects they use (computer keys and computer mice). My analysis asks specifically how do men and women's skin-associated bacterial communities from this study differ? How are they similar? I hypothesize they will differ most in which bacterial communities they have and how long the sequence lengths will be. 

Sources:
https://www.ncbi.nlm.nih.gov/pubmed/27551280
https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?study=ERP022657

# Methods

## Sample origin and sequencing

Nine women and men of equivalent age and health (four women, five men) were used for the computer mouse study, all of whom worked in the same building at the University of Colorado. The computer mouse belonging to the individual and the palm of the individual's dominant hand were swabbed. Each computer mouse had been touched by the individual in the last 12 hours. Before sampling, all individuals practiced their regular hygeine routines. They compared the swabs from the nine individuals to swabs from 270 other hands from healthy individuals ranging from 18-40 years old. In each sample, the 16S rRNA genes were amplified then carried out via PCR. If sequences were less than 200 or more than 300 base pairs in length, they were excluded from the study. 

## Computational

First, the sequences were obtained from the NCBI database which is open for the public. The files were downloaded to a raw data directory. Next, a QC (quality check) report was run on each individual file. The QC reports are helpful to see how viable each sequence is for analysis; if the majority of the sequence is above a 30 score, then we consider it to be sequenced well. Typically, the 3' end will deteriorate, which is normal. Once the files were checked for quality, they were trimmed using Trimmomatic, which discards any sequence less than 150 base pairs or a base score below 25, which is similar to the experiment's desire to exlude any lengths less than 200 base pairs. Once trimmed, the files were converted to fasta files (from fastq) because BLAST will only work on fasta files. Lastly, the fasta files were put through a BLAST analysis. 

# Results

```{r load-libraries, message = FALSE}
# Be sure to install these packages before running this script
# They can be installed either with the intall.packages() function
# or with the 'Packages' pane in RStudio

# load packages
library("dplyr")
library("tidyr")
library("knitr")
library("ggplot2")
```

```{r make-read-in-data-function, echo=FALSE}
# Output format from BLAST is as detailed on:
# https://www.ncbi.nlm.nih.gov/books/NBK279675/
# In this case, we used: '10 sscinames std'
# 10 means csv format
# sscinames means unique Subject Scientific Name(s), separated by a ';'
# std means the standard set of result columns, which are:
# 'qseqid sseqid pident length mismatch
# gapopen qstart qend sstart send evalue bitscore',


# this function takes as input a quoted path to a BLAST result file
# and produces as output a dataframe with proper column headers
# and the 'qseqid' column split into sample and seq number
read_blast_output <- function(filename) {
  data_in <- read.csv(filename,
                      header = FALSE, # files don't have column names in them
                      col.names = c("sscinames", # unique Subject Sci Name(s)
                                    "qseqid",    # Query Seq-id
                                    "sseqid",    # Subject Seq-id
                                    "pident",    # Percntge of identical matches
                                    "length",    # Alignment length
                                    "mismatch",  # Number of mismatches
                                    "gapopen",   # Number of gap openings
                                    "qstart",    # Start of alignment in query
                                    "qend",      # End of alignment in query
                                    "sstart",    # Start of alignment in subj
                                    "send",      # End of alignment in subject
                                    "evalue",    # Expect value
                                    "bitscore"))  # Bit score

  # Next we want to split the query sequence ID into
  # Sample and Number components so we can group by sample
  # They originally look like "ERR1942280.1"
  # and we want to split that into two columns: "ERR1942280" and "1"
  # we can use the separate() function from the tidyr library to do this
  # Note that we have to double escape the period for this to work
  # the syntax is
  # separate(column_to_separate,
  # c("New_column_name_1", "New_column_name_2"),
  # "seperator")
  data_in <- data_in %>%
    separate(qseqid, c("sample_name", "sample_number"), "\\.")
}
```

```{r read-in-BLAST-data, echo=FALSE}
# this makes a vector of all the BLAST output file names, including
# the name(s) of the directories they are in
files_to_read_in <- list.files(path = "output/blast",
                               full.names = TRUE)

# We need to create an empty matrix with the right number of columns
# so that we can rbind() each dataset on to it
joined_blast_data <- matrix(nrow = 0,
                            ncol = 14)

# now we loop over each of the files in the list and append them
# to the bottom of the 'joined_blast_data' object
# we do this with the rbind() function and the function we
# made earlier to read in the files, read_blast_output()
for (filename in files_to_read_in) {
  joined_blast_data <- rbind(joined_blast_data,
                             read_blast_output(filename))
}
```

```{r read-in-metadata-and-join, echo=FALSE}
# Next we want to read in the metadata file so we can add that in too
# This is not a csv file, so we have to use a slightly different syntax
# here the `sep = "\t"` tells the function that the data are tab-delimited
# and the `stringsAsFactors = FALSE` tells it not to assume that things are
# categorical variables
metadata_in <- read.table(paste0("data/metadata/",
                                 "fierer_forensic_hand_mouse_SraRunTable.txt"),
                          sep = "\t",
                          header = TRUE,
                          stringsAsFactors = FALSE)

# Finally we use the left_join() function from dplyr to merge or 'join' the
# combined data and metadata into one big table, so it's easier to work with
# in R the `by = c("Run_s" = "sample_name")` syntax tells R which columns
# to match up when joining the datasets together
joined_blast_data_metadata <- metadata_in %>%
  left_join(joined_blast_data,
            by = c("Run_s" = "sample_name"))
```


```{r histograms-for-female-and-male-pident, echo=FALSE}
# Here we're using the dply piping syntax to select a subset of rows matching a
# criteria we specify (using the filter) function, and then pull out a column
# from the data to make a histogram. We don't need to tell the hist() function
# which data to use, because that's piped in, but we do have to give the
# hist() function the title and axis label we'd like to use for the figure

# Histograms for basic percent identity for females versus males.
joined_blast_data_metadata %>%
  filter(sex_s == "female") %>%
  filter(env_material_s == "sebum") %>%
  pull(pident) %>%
  hist(main = "Percent Identity Skin Samples in Females",
       xlab = "Percent")
joined_blast_data_metadata %>%
  filter(sex_s == "male") %>%
  filter(env_material_s == "sebum") %>%
  pull(pident) %>%
  hist(main = "Percent Identity Skin Samples in Males",
       xlab = "Percent")
```

In **Figures 1 and 2**, we see the overall percent identity match frequency for female and male skin samples.

### Top 3 Species Shown in Females

```{r tables-top3-females}
joined_blast_data_metadata %>%
  filter(sex_s == "female") %>%
  filter(env_material_s == "sebum") %>%
  group_by(sscinames) %>%
  count() %>%
  arrange(desc(n)) %>%
  head(3) %>%
  kable()
```

**Table 1**: Shown here are the top three bacterial species found on females in the study. 

### Top 3 Species Shown in Males

```{r tables-top3-males}
joined_blast_data_metadata %>%
  filter(sex_s == "male") %>%
  filter(env_material_s == "sebum") %>%
  group_by(sscinames) %>%
  count() %>%
  arrange(desc(n)) %>%
  head(3) %>%
  kable()
```

**Table 2**: Shown here are the top three bacterial species found on males in the study. 

### Individual Mismatch Comparison 

```{r ggplots-individual-mismatch}
ggplot(joined_blast_data_metadata ,
  aes(x = mismatch,
      fill = sample_type_s)) +
  geom_density(alpha=0.8) +
facet_wrap(~host_subject_id_s)
```

**Figure 3**: Shown in the figure is every individuals' mismatch rate for both computer mouse data and right palm data. 

### Individual Length Comparison

```{r ggplots-individual-length}
ggplot(joined_blast_data_metadata ,
  aes(x = length,
      fill = sample_type_s)) +
  geom_density(alpha=0.8) +
facet_wrap(~host_subject_id_s)

```

**Figure 4**: Shown in the figure are comparable sequence lengths between all individuals from both computer mouse data and right palm data. 

# Discussion

The data from the Fierer et al. study are confusing yet interesting. For example, in Figures 1 and 2, we found that the female percent identity match was noticeably much larger than that of males. For females, the identity match was typically above 95%, while in males it was usually 85%. The reason why females have a high percent identity? There's many unsupported reasons why this is the case. Potentially it has something to do with hand washing techniques, or something as simple as the way the swabs were done.

To delve further, I examined the top three bacterial species which showed up in females and males, as there wasn't too much overlap. For females, the most common was Bartonella washoensis with 678 matches. This particular species of bacteria is known to cause meningitis in humans and is often passed virally. However, other sources indicate that its pathophysiology is more commonly related to myocarditis (inflammation of the heart muscle) which is rare in women. The research was confounding until I looked at the second most common bacteria found in women, which was Acidovorax sp (173 occurances). While that species is mainly known for infecting plants, it is also known for causing meningitis, wound infections, and UTIs. The appearance of two bacterial species causing meningitis in the women samples is interesting, but I'm hesitant to make any broad claims based on this limited data on only four female subjects. In the male samples, the top two bacterial species were found either from gills of gutless awning-clams or places with lakes, wetlands, or mud. While the connection is interesting, I don't find it as helpful as the last most common bacteria found, which was again Acidovorax sp. Clearly, within the University of Colorado building there is a common appearance of this particular bacteria found on both female and male hands. 

Lastly, I wanted to do a broad comparison between all experiment participants in terms of their sequence mismatch and length. In Figure 3, F7 shows the highest rate of mismatches. Noticeably, men have an even spread of very little mismatches, while females show a visible skew to the left. What is also visible is that when men did have mismatches, it was from the swab done on the right palm, while women primarily had mismatches from the computer mouse data. This could either be experimental error or this could indicate **why** we see such a big difference in percent identity between males and females from Figures 1 & 2. In Figure 4, we see a common skew to the right for both females and males, except arguably subjects M1. We also see the sequence length is typically larger when swabbed from the palm versus the computer mouse, which is interesting, but no well-developed hypotheses about why can be formed at this time.

Overall, future steps this study could take would be swabbing *more* individuals and using other devices touched by others rather than a personal computer. I would argue, however, that studies like these do not indicate that such techniques provide ample support for forensic research because, in real life, there are no controls which you can compare your sequences off of. Unless the forensic team was biased and already had a suspect in mind. Nevertheless, the results of the study are interesting and more can be learned about skin-associated bacterial communities from studies such as these.

```{r summary-table}
# Finally, we'd like to be able to make a summary table of the counts of
# sequences for each taxa for each sample. To do that we can use the table()
# function. We add the kable() function as well (from the tidyr package)
# in order to format the table nicely when the document is knitted
kable(table(joined_blast_data_metadata$sscinames,
            joined_blast_data_metadata$Run_s))
            
```


